libslide/src/bignum/sub.rs

```diff
@@ -0,0 +1,170 @@
+            if !(self.is_neg || other.is_neg) {
+                is_neg = true;
+            }
+            lhs = self;
+            rhs = other;
+        }
+
+        // make the decimal vectors the same size
+        // fill lhs and rhs vectors as well
+        let self_size: usize = lhs.dec.len();
+        let other_size: usize = rhs.dec.len();
+        let (res_int, lhs_int, res_dec, lhs_dec) = if self_size > other_size {
+            let mut equal_dec = vec![0; self_size];
+            equal_dec[..other_size].clone_from_slice(&rhs.dec[..other_size]);
+            (rhs.int, lhs.int, equal_dec, lhs.dec)
+        } else {
+            let mut equal_dec = vec![0; other_size];
+            equal_dec[..self_size].clone_from_slice(&lhs.dec[..self_size]);
+            (rhs.int, lhs.int, rhs.dec, equal_dec)
+        };
+
+        // now we need to make the bigger number the res dec
+        // 1. Decimal
+
+        let mut res: i8;
+        let mut carry: i8 = 0;
+        let mut res_dec_asi8 = recast_vec::<u8, i8>(res_dec);
+        let lhs_dec_asi8 = recast_vec::<u8, i8>(lhs_dec);
+        let mut res_int_asi8 = recast_vec::<u8, i8>(res_int);
+        let lhs_int_asi8 = recast_vec::<u8, i8>(lhs_int);
+
+        for i in (0..res_dec_asi8.len()).rev() {
+            res = res_dec_asi8[i] - lhs_dec_asi8[i] - carry;
+            carry = (res < 0) as i8;
+            res += carry * 10;
+            res_dec_asi8[i] = res;
+        }
+
+        // 2. Integer
+        for i in 0..lhs_int_asi8.len() {
+            res = res_int_asi8[i] - lhs_int_asi8[i] - carry;
+            carry = (res < 0) as i8;
+            res += carry * 10;
+            res_int_asi8[i] = res;
+        }
+
+        // Propogate the rest of the carry
+        let mut i = lhs_int_asi8.len();
+        while i < res_int_asi8.len() {
+            res = res_int_asi8[i] as i8 - carry;
+            carry = (res < 0) as i8;
+            res += carry * 10;
+            res_int_asi8[i] = res;
+            i += 1;
+        }
+
+        // remove preceeding zeros
+        let mut count = 0;
+        for i in (0..res_int_asi8.len()).rev() {
+            if res_int_asi8[i] == 0 {
+                count += 1;
+            } else {
+                break;
+            }
+        }
+
+        res_int_asi8.truncate(res_int_asi8.len() - count);
+
+        count = 0;
+        for i in (0..res_dec_asi8.len()).rev() {
+            if res_dec_asi8[i] == 0 {
+                count += 1;
+            } else {
+                break;
+            }
+        }
+        res_dec_asi8.truncate(res_dec_asi8.len() - count);
+
+        // quick optimization to remove -0 from possible results.
+        if is_neg && res_int_asi8.is_empty() && res_dec_asi8.is_empty() {
+            is_neg = false;
+        }
+
+        Bignum {
+            is_neg,
+            dec: recast_vec::<i8, u8>(res_dec_asi8),
+            int: recast_vec::<i8, u8>(res_int_asi8),
```

Comment:
I think you can elide the type parameters

Reply:
how? doesn't it need to know what type to cast to?

Reply:
Try it